### PR TITLE
Tbrands 27 algolia assortment content source

### PR DESCRIPTION
**Pull request by JackHowa**
## Description

Add algolia content source

## Jira Ticket

- [TMEDIA-27](https://arcpublishing.atlassian.net/browse/TMEDIA-27)

## Acceptance Criteria

A new block in the arc-themes-blocks repo - algolia-assortment-content-source

A Fusion Content Source that can query the Algolia Search API with the following parameters

Index

Context - Algolia Param - ruleContexts

Filters - Algolia Param - filters

Query - Algolia Param - query

Size - Algolia Param - hitsPerPage - default = 20

Page - Algolia Param - page - default = 0

The content source relies on ApplicationId and APIKey to set up the Algolia search methods

Content Source has TTL set to the Fusion minimum of 2 minutes

## Test Steps

1. Go to feature pack repo 
2. Checkout this branch `git checkout TBRANDS-27-algolia-assortment-content-source}`
3. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/algolia-assortment-content-source-block`
4. Go to blocks repo
5. Checkout this branch `git checkout TBRANDS-27-algolia-assortment-content-source}`
6. Go to feature pack repo.  https://github.com/WPMedia/arc-themes-feature-pack/pull/218
7. Ensure your .env has plaintext values for `algoliaSearchKey` and `algoliaAppId`. See https://github.com/WPMedia/arc-themes-feature-pack/pull/209#issue-1154629794 for getting these values if you don't have them already
8. Make sure you've added `@local` to the block name in the blocks.json
9. Run `npx fusion start -f -l @wpmedia/algolia-assortment-content-source-block`
10. Go to content debugger on local for algolia assortment http://localhost/pagebuilder/tools/debugger?_website=arc-commerce&type=content-debugger&source=algolia-assortment
11. Run a query like and notice the first image url. Notice there are 20 items returned. This means there are 20 items by default and it starts on page 0. `searchdemoproduct_commercedev_` for index, `shoes` for query, `featured_shoes` for rule contexts
Screen Shot 2022-04-12 at 10 57 59
12. Now set the same query to 21 count. You'll see 21 items
13. Now set the page to 1. You will see a different image url than by default before
Screen Shot 2022-04-12 at 10 59 51
14. Now add a filter to notice the effect of adding a specific sku. (In our example, there's 3 items with the same sku.) The filter text should be: `("sku": "SHOE34YXGQYWM45Y")`
Screen Shot 2022-04-12 at 11 01 49

### Before

- No algolia assortment content source

### After

- have an algolia assortment content source! 

## Dependencies or Side Effects

- Need the environment files. I've added them in the feature pack pr for the dev commerce environment with the key encrypted
- Merge and deploy https://github.com/WPMedia/arc-themes-feature-pack/pull/218

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
        please explain why (so that we can fix it whenever it gets refactored).
- [x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

### Notes 
- Go to https://github.com/WPMedia/arc-themes-feature-pack/pull/209#issue-1154629794 for getting the algoliaSearchKey set in `.env`. The `algoliaAppId` should also be set in the `.env` 
- Add     "@wpmedia/algolia-assortment-content-source-block@local" to blocks.json. run `npx fusion start -f -l  @wpmedia/algolia-assortment-content-source-block` to get this working locally
- I've added additional comments after feedback from @marissa-halpert 

Working example with options set from ticket 

index
searchdemoproduct_commercedev_

Screen Shot 2022-04-11 at 14 05 05

### Questions
- I don't seem to be able to use filters. I've tried various configurations https://www.algolia.com/doc/api-reference/api-parameters/filters/ , including is_fk_advantage_product:false and  "is_fk_advantage_product":"false". Is there something that I'm missing with the format or the data? Looking for help on this end as I did not see it in Nick's pr 
- For the algolia app id, I set this in the environment file. I'm assuming this is ok based on the ticket ac https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TBRANDS-27 Is this ok? 

### Todos

- [x] Update readme with basic info and params
- [x] Create feature pack pr with the environment values, including encrypted search key and plaintext app id
- [x] Add tests with mocking algolia api

### CODE REVIEW

**Comment by matthewroach** [comment]
@JackHowa - This is working for me

Microsoft Edge-2022-04-11-22-05 28

**Comment by JackHowa** [comment]
> @JackHowa - This is working for me
> 
> Microsoft Edge-2022-04-11-22-05 28

huh cool. I'll work on this and get it ready in a bit. thanks!

**Comment by matthewroach** [comment]
@JackHowa - Can you update the package files to ensure the actions are not failing please

File: _templates/content-source/new/default-test.ejs.t [hunk 1]
```diff
@@ -4,7 +4,7 @@ to: blocks/<%= h.inflection.dasherize(block_name) %>-content-source-block/source
 ---
 import contentSource from './<%= h.inflection.dasherize(block_name) %>';
 
-describe('Test <%= h.changeCase.title(content_source_name) %> content source', () => {
+describe('Test <%= h.changeCase.title(block_name) %> content source', () => {
```
Comment (JackHowa) [inline _templates/content-source/new/default-test.ejs.t:7]: noticed issue with this naming and fixed

**Comment by JackHowa** [comment]
> @JackHowa - Can you update the package files to ensure the actions are not failing please

@matthewroach updated

File: blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	const { filters, index, query, ruleContexts } = contentSourceParameters;
+	let { hitsPerPage, page } = contentSourceParameters;
+
+	// parameters in content source are defined as null by default
+	if (hitsPerPage === null) {
+		hitsPerPage = 20;
+	}
+	if (page === null) {
+		page = 0;
+	}
```
Comment (matthewroach) [inline blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:23]: Any reason not to set defaults like so ?

```suggestion
	const {
		filters,
		index,
		query,
		ruleContexts,
		hitsPerPage = 20,
		page = 0,
	} = contentSourceParameters;
```

Comment (JackHowa) [reply blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:23]: yes. in the comment and test, I said the values were coming in as `null`. So we can't use default parameters @matthewroach

File: blocks/algolia-assortment-content-source-block/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# @wpmedia/algolia-assortment-content-source-block
+
+The Algolia assortment content source block allows you to display products from an Algolia index.
+
+## ANS Schema
```
Comment (matthewroach) [inline blocks/algolia-assortment-content-source-block/README.md:5]: Minor - but algolia is not using ANS Schema

Comment (JackHowa) [reply blocks/algolia-assortment-content-source-block/README.md:5]: ah yeah. one sec

Comment (JackHowa) [reply blocks/algolia-assortment-content-source-block/README.md:5]: updated @matthewroach

**Review by matthewroach — APPROVED** [review]

File: blocks/algolia-assortment-content-source-block/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+The Algolia assortment content source block allows you to display products from an Algolia index. See Algolia docs for more content info: https://www.algolia.com/doc/guides/sending-and-managing-data/prepare-your-data/#algolia-records
+
+## ANS Schema
+
+not applicable
+
+## Configurable Params
+
+```
+    filters: "text",
```
Comment (brentmiller1973) [inline blocks/algolia-assortment-content-source-block/README.md:12]: Wondering if we could expand on the params a bit here?  More description and examples possibly?

File: blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+		error.statusCode = 500;
+		throw error;
+	}
+};
+export default {
+	fetch,
+	params,
+	transform: (data) => data,
+	// Time to live cache set to the Fusion minimum of 2 minutes (120 seconds)
+	ttl: 120,
```
Comment (ecarlisle) [inline blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:48]: @JackHowa : FYI as a topic to revisit but not a blocker. The Commerce team has mentioned being amenable to a 5-minute TTL for assortment data. Delivery may appreciate that as well.

File: blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	const clientSearchIndex = client.initIndex(index);
+	try {
+		const { hits } = await clientSearchIndex.search(query, {
+			// array of strings
+			// https://www.algolia.com/doc/api-reference/api-parameters/ruleContexts/
+			ruleContexts: [ruleContexts],
+			filters,
+			hitsPerPage,
+			page,
+		});
```
Comment (ecarlisle) [inline blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:35]: Also for another day, but when the Commerce schema is final we can use this as a filter to minimize payload:
https://www.algolia.com/doc/api-reference/api-parameters/attributesToRetrieve/

**Review by ecarlisle — APPROVED** [review]
Just added some notes for future but I think we're in a great place!

File: blocks/algolia-assortment-content-source-block/README.md (hunk 4 above)
Comment (JackHowa) [reply blocks/algolia-assortment-content-source-block/README.md:12]: updated with some description and links to algolia's documentation. I don't want to recreate their docs too much

Comment (JackHowa) [reply blocks/algolia-assortment-content-source-block/README.md:12]: Updated! thanks for the feedback @visionbegin. I added some links to algolia's documentation related. I don't want to recreate their docs too much (they have really great docs with examples).

File: blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js (hunk 5 above)
Comment (JackHowa) [reply blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:48]: gotcha. I was going off of the ticket requirements. if that's not the goal, then I can change this @ecarlisle

File: blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js (hunk 6 above)
Comment (JackHowa) [reply blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:35]: Yep good point. Is this ok to resolve? @ecarlisle

File: blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js (hunk 5 above)
Comment (matthewroach) [reply blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:48]: @ecarlisle Where was this mentioned? As we refined the ticket with the 2 minute ttl - I want to make sure we are all on the same page with decisions being made.
We can move forward with the 2 minute ttl for now, and create a ticket to update it to something else if needed

Comment (ecarlisle) [reply blocks/algolia-assortment-content-source-block/sources/algolia-assortment.js:48]: @matthewroach : Long ago during the initial consultations Algolia gave us. I agree with starting with 2 minutes and making a ticket for any future changes.

File: blocks/algolia-assortment-content-source-block/README.md (hunk 4 above)
Comment (brentmiller1973) [reply blocks/algolia-assortment-content-source-block/README.md:12]: That works for me!

Comment (brentmiller1973) [reply blocks/algolia-assortment-content-source-block/README.md:12]: Issue is resolved

**Review by brentmiller1973 — APPROVED** [review]

**Pull request merged**